include/nbl/builtin/hlsl/complex.hlsl

```diff
@@ -0,0 +1,346 @@
+// Could X-Macro, left as-is for readability
+// These all have to be specialized because of the identity that can't be initialized inside the struct definition, 
+// and the latter (mul, div) have two overloads for operator()
+
+template<typename Scalar> 
+struct plus< complex_t<Scalar> > 
+{
+    complex_t<Scalar> operator()(NBL_CONST_REF_ARG(complex_t<Scalar>) lhs, NBL_CONST_REF_ARG(complex_t<Scalar>) rhs) 
+    {
+        return lhs + rhs;                                                             
+    }
+
+    NBL_CONSTEXPR_STATIC_INLINE complex_t<Scalar> identity;                                          
+};
+
+template<typename Scalar> 
+struct minus< complex_t<Scalar> > 
+{
+    complex_t<Scalar> operator()(NBL_CONST_REF_ARG(complex_t<Scalar>) lhs, NBL_CONST_REF_ARG(complex_t<Scalar>) rhs) 
+    {
+        return lhs - rhs;                                                             
+    }
+
+    NBL_CONSTEXPR_STATIC_INLINE complex_t<Scalar> identity;                                          
+};
+
+template<typename Scalar> 
+struct multiplies< complex_t<Scalar> > 
+{
+    complex_t<Scalar> operator()(NBL_CONST_REF_ARG(complex_t<Scalar>) lhs, NBL_CONST_REF_ARG(complex_t<Scalar>) rhs) 
+    {
+        return lhs * rhs;                                                             
+    }
+
+    complex_t<Scalar> operator()(NBL_CONST_REF_ARG(complex_t<Scalar>) lhs, NBL_CONST_REF_ARG(Scalar) rhs) 
+    {
+        return lhs * rhs;                                                             
+    }
+
+    NBL_CONSTEXPR_STATIC_INLINE complex_t<Scalar> identity;                                          
+};
+
+template<typename Scalar> 
+struct divides< complex_t<Scalar> > 
+{
+    complex_t<Scalar> operator()(NBL_CONST_REF_ARG(complex_t<Scalar>) lhs, NBL_CONST_REF_ARG(complex_t<Scalar>) rhs) 
+    {
+        return lhs / rhs;                                                             
+    }
+
+    complex_t<Scalar> operator()(NBL_CONST_REF_ARG(complex_t<Scalar>) lhs, NBL_CONST_REF_ARG(Scalar) rhs) 
+    {
+        return lhs / rhs;                                                             
+    }
+
+    NBL_CONSTEXPR_STATIC_INLINE complex_t<Scalar> identity;                                          
+};
+
+
+// Out of line generic initialization of static member data not yet supported so we X-Macro identities for Scalar types we want to support
+// (left X-Macro here since it's pretty readable)
+
+#define COMPLEX_ARITHMETIC_IDENTITIES(SCALAR) \
+template<> \
+const static complex_t< SCALAR > plus< complex_t< SCALAR > >::identity = { SCALAR (0), SCALAR (0)}; \
+template<> \
+const static complex_t< SCALAR > minus< complex_t< SCALAR > >::identity = { SCALAR (0),  SCALAR (0)}; \
+template<> \
+const static complex_t< SCALAR > multiplies< complex_t< SCALAR > >::identity = { SCALAR (1),  SCALAR (0)}; \
+template<> \
+const static complex_t< SCALAR > divides< complex_t< SCALAR > >::identity = { SCALAR (1),  SCALAR (0)};
```

Comment:
you need to use `promote` from to get your 1 to expand to 1,1,... as neeeded

Reply:
Where is `promote` declared?